devices/tty-bios/tty-bios.c

```diff
@@ -0,0 +1,499 @@
+	"\033[A",  /* ekey_up */
+	"\033[B",  /* ekey_down */
+	"\033[C",  /* ekey_right */
+	"\033[D",  /* ekey_left */
+	"\033[3~", /* ekey_delete */
+};
+
+
+/* ANSI code to VGA foreground character attribute conversion table */
+static const unsigned char ansi2fg[] = {
+	0x00, 0x04, 0x02, 0x06,
+	0x01, 0x05, 0x03, 0x07
+};
+
+
+/* ANSI code to VGA background character attribute conversion table */
+static const unsigned char ansi2bg[] = {
+	0x00, 0x40, 0x20, 0x60,
+	0x10, 0x50, 0x30, 0x70
+};
+
+
+typedef struct {
+	unsigned char rows;      /* Terminal height */
+	unsigned char cols;      /* Terminal width */
+	unsigned char attr;      /* Character attribute */
+	unsigned char esc;       /* Escape sequence state */
+	unsigned char parmi;     /* Escape sequence parameter index */
+	unsigned char parms[10]; /* Escape sequence parameters buffer */
+	const char *ekey;        /* Extended key code to send */
+} ttybios_t;
+
+
+struct {
+	ttybios_t ttys[TTYBIOS_MAX_CNT];
+} ttybios_common;
+
+
+/* Retrieves cursor position */
+static void ttybios_getCursor(unsigned char *row, unsigned char *col)
+{
+	unsigned short pos;
+
+	__asm__ volatile(
+		"xorb %%bh, %%bh; "
+		"movb $0x3, %%ah; "
+		"pushl $0x10; "
+		"pushl $0x0; "
+		"pushl $0x0; "
+		"call _interrupts_bios; "
+		"addl $0xc, %%esp; "
+	: "=d" (pos)
+	:: "eax", "ebx", "ecx", "memory", "cc");
+
+	*row = pos >> 8;
+	*col = pos;
+}
+
+
+/* Sets cursor position */
+static void ttybios_setCursor(unsigned char row, unsigned char col)
+{
+	__asm__ volatile(
+		"xorb %%bh, %%bh; "
+		"movb $0x2, %%ah; "
+		"pushl $0x10; "
+		"pushl $0x0; "
+		"pushl $0x0; "
+		"call _interrupts_bios; "
+		"addl $0xc, %%esp; "
+	:: "d" (((unsigned short)row << 8) | col)
+	: "eax", "ebx", "memory", "cc");
+}
+
+
+/* Checks for available characters at terminal input */
+static int ttybios_checkc(void)
+{
+	int ret;
+
+	__asm__ volatile(
+		"movb $0x1, %%ah; "
+		"pushl $0x16; "
+		"pushl $0x0; "
+		"pushl $0x0; "
+		"call _interrupts_bios; "
+		"jnz 0f; "
+		"xorl %%eax, %%eax; "
+		"jmp 1f; "
+		"0: "
+		"movl $0x1, %%eax; "
+		"1: "
+		"addl $0xc, %%esp; "
+	: "=a" (ret)
+	:: "memory", "cc");
+
+	return ret;
+}
+
+
+/* Reads character and its scan code from terminal input */
+static void ttybios_getc(char *sc, char *c)
+{
+	unsigned short key;
+
+	__asm__ volatile(
+		"xorb %%ah, %%ah; "
+		"pushl $0x16; "
+		"pushl $0x0; "
+		"pushl $0x0; "
+		"call _interrupts_bios; "
+		"addl $0xc, %%esp; "
+	: "=a" (key)
+	:: "memory", "cc");
+
+	*sc = key >> 8;
+	*c = key;
+}
+
+
+/* Writes character to terminal output */
+static void ttybios_putc(unsigned char attr, char c)
+{
+	__asm__ volatile(
+		"pushl $0x10; "
+		"pushl $0x0; "
+		"pushl $0x0; "
+		"call _interrupts_bios; "
+		"addl $0xc, %%esp; "
+	:: "a" (0xe00 | c), "b" (attr)
+	: "memory", "cc");
+}
+
+
+/* Returns terminal instance */
+static ttybios_t *ttybios_get(unsigned int minor)
+{
+	if (minor >= TTYBIOS_MAX_CNT)
+		return NULL;
+
+	return &ttybios_common.ttys[minor];
```

Comment:
It's not a problem for now since we have only one VGA-keyboard based terminal device (TTYBIOS_MAX_CNT is 1) and it's always active. Proper checks for active/disabled device is something that I'd like to add in another PR.